src/ruby/ext/grpc/rb_channel.c

```diff
@@ -261,6 +263,65 @@ static VALUE grpc_rb_channel_init(int argc, VALUE *argv, VALUE self) {
   return self;
 }
 
+/*
+  call-seq:
+    ch = Channel.insecure_create_from_fd("fd:42", 42, {'arg1': 'value1'})
+
+  Creates channel instances from an existing file descriptor. */
+static VALUE grpc_rb_channel_insecure_create_from_fd(
+    VALUE klass, VALUE rb_target, VALUE rb_fd, VALUE rb_args) {
+#ifdef GPR_SUPPORT_CHANNELS_FROM_FD
+  VALUE self;
+  grpc_rb_channel *wrapper = NULL;
+  grpc_channel *ch = NULL;
+  grpc_channel_args args;
+  channel_init_try_register_stack stack;
+  int stop_waiting_for_thread_start = 0;
+  MEMZERO(&args, grpc_channel_args, 1);
+
+  grpc_ruby_once_init();
+  rb_thread_call_without_gvl(
+      wait_until_channel_polling_thread_started_no_gil,
+      &stop_waiting_for_thread_start,
+      wait_until_channel_polling_thread_started_unblocking_func,
+      &stop_waiting_for_thread_start);
+
+  if (TYPE(rb_fd) != T_FIXNUM) {
+    rb_raise(rb_eTypeError, "bad fd, wanted an integer");
+    return Qnil;
+  }
+
+  grpc_rb_hash_convert_to_channel_args(rb_args, &args);
+
+  self = grpc_rb_channel_alloc(klass);
+  TypedData_Get_Struct(self, grpc_rb_channel, &grpc_channel_data_type, wrapper);
+  ch = grpc_insecure_channel_create_from_fd(StringValueCStr(rb_target),
```

Comment:
Can we de-dup this code by changing around the channel constructor by changing around this code to take a function pointer, and then e.g. this `create_from_fd` constructor can perhaps pass in `grpc_insecure_channel_create_from_fd` to use to create the actual channel?